### PR TITLE
MRG: Add oversampled temporal projection

### PR DESCRIPTION
**Pull request by larsoner**
Finally adds the [OTP algorithm](https://dx.doi.org/10.1109/TBME.2017.2734641), useful for denoising HF responses, (some) bad channel repair, and raw data browsing.

### CODE REVIEW

**Comment by larsoner** [comment]
https://6732-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.wcSIsWq/html/auto_examples/preprocessing/plot_otp.html

File: examples/preprocessing/plot_otp.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+"""
+===========================================================
+Plot sensor denoising using oversampled temporal projection
+===========================================================
+
+This demonstrates denoising using the OTP algorithm on a dataset
+with noise arising predominantly from sensors (not the environment).
```
Comment (agramfort) [inline examples/preprocessing/plot_otp.py]: please add a ref to paper here.

File: mne/preprocessing/utils.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    step = nperseg - noverlap
+    binsums = np.sum((win[ii * step:(ii + 1) * step]
+                      for ii in range(nperseg // step)), axis=0)
+    if nperseg % step != 0:
+        binsums[:nperseg % step] += win[-(nperseg % step):]
+    deviation = binsums - np.median(binsums)
+    return np.max(np.abs(deviation)) < tol
+
+
+def _get_lims_COLA(n_samp, n_times, sfreq, picks):
```
Comment (agramfort) [inline mne/preprocessing/utils.py]: what is this _get_lims_COLA ?

also it should be _get_lims_cola

we don't put capitals in function names.

File: mne/preprocessing/otp.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        if not np.isfinite(data_picked).all():
+            raise RuntimeError('non-finite data (inf or nan) found in raw '
+                               'instance')
+        # demean our slice and our copy
+        data_picked_means = np.mean(data_picked, axis=-1, keepdims=True)
+        data_picked -= data_picked_means
+        # scale the copy that will be used to form the temporal basis vectors
+        # so that _orth_svdvals thresholding should work properly with
+        # different channel types (e.g., M-EEG)
+        norms = _normalize_vectors(data_picked)
```
Comment (agramfort) [inline mne/preprocessing/otp.py]: can you compute norms 3liens below as you don't use them before the for ?

File: mne/preprocessing/tests/test_otp.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+# Authors: Eric Larson <larson.eric.d@gmail.com>
+#
+# License: BSD 3 clause
+
+import os.path as op
+
+import pytest
+
+import numpy as np
+from numpy.fft import rfft, rfftfreq
```
Comment (agramfort) [inline mne/preprocessing/tests/test_otp.py]: why not taking it from scipy ?

File: mne/preprocessing/utils.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# License: BSD (3-clause)
+
+import numpy as np
+
+from ..utils import logger
+
+
+# adapted from SciPy
+
+def check_COLA(win, nperseg, noverlap, tol=1e-10):
```
Comment (agramfort) [inline mne/preprocessing/utils.py]: check_COLA -> check_cola or better

_check_overlap_add_constraint (private)

Comment (larsoner) [reply mne/preprocessing/utils.py]: I'd rather keep it this way to mirror SciPy more closely / make translations back and forth easier

File: mne/preprocessing/tests/test_otp.py (hunk 4 above)
Comment (larsoner) [reply mne/preprocessing/tests/test_otp.py]: SciPy's RFFT returns data in a different, more annoying form

File: mne/preprocessing/utils.py (hunk 2 above)
Comment (larsoner) [reply mne/preprocessing/utils.py]: acronyms should be capitalized, it's the "Constant OverLap Add" constraint

**Comment by agramfort** [comment]
you have a big flux jump in sample on channel MEG 2313 at about t=65s

that could be a more convincing dataset for the example.

File: mne/preprocessing/utils.py (hunk 5 above)
Comment (agramfort) [reply mne/preprocessing/utils.py]: k

**Comment by agramfort** [comment]
we never capitatlize acronyms in our functions or parameters. for example
lcmv, dics, snr etc. unless they are constant. it's the coding semantic.

File: mne/preprocessing/otp.py (hunk 3 above)
Comment (larsoner) [reply mne/preprocessing/otp.py]: no, `_normalize_vectors` operates inplace so they are actually used (already divided out by the call itself)

**Comment by larsoner** [comment]
Comments addressed @agramfort, changed the example to use the `bst_phantom_elekta` dataset because it has flux jumps and you can see localization improvement

**Comment by larsoner** [comment]
https://6790-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.eNuqDmi/html/auto_examples/preprocessing/plot_otp.html

**Comment by larsoner** [comment]
@mmagnuski happy too?

**Comment by mmagnuski** [comment]
I'm happy you are adding this and feel confident it works flawlessly, but I won't have time to look into the details this week (but given you are the author of the paper you implement I'm not sure this is necessary). I looked through the example and I like it a lot.

**Comment by larsoner** [comment]
Thanks for looking @mmagnuski @agramfort. Given that feedback and existing reviews (including from previous PR) I'll go ahead and merge once the rebase comes back green

**Comment by agramfort** [comment]
+1

**Pull request merged**